iwyu.cc

```diff
@@ -2224,6 +2224,27 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
+  bool VisitTypeTraitExpr(clang::TypeTraitExpr* expr) {
+    if (CanIgnoreCurrentASTNode())  return true;
+
+    // We assume that all type traits with >= 2 arguments require
+    // full type information even for pointer types. For example,
+    // this is the case for `__is_convertible_to` trait.
+    if (expr == nullptr || expr->getNumArgs() <= 1)  return true;
```

Comment:
It's unfortunate that the `CanIgnoreCurrentASTNode()` checks everywhere are packed on a single line, because I think that's really poor style (it's much more ergonomic to use debuggers when return stmts are on their own lines).

Also `getNumArgs() < 2` makes it clearer that 2 is the magic number.